### PR TITLE
[pytorch] [build, test] [neuron]:dlc for runtime1.1

### PR DESCRIPTION
**Pull request by aws-vrnatham**
Signed-off-by: Venky Natham <vrnatham@amazon.com>

*Issue #, if available:*

## PR Checklist
- [ x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON Checklist
* When creating a PR:
- [ x] I've modified `src/config/build_config.py` in my PR branch by setting `ENABLE_EI_MODE = True` or `ENABLE_NEURON_MODE = True`
* When PR is reviewed and ready to be merged:
- [ ] I've reverted the code change on the config file mentioned above
#### Benchmark Checklist
* When creating a PR:
- [ ] I've modified `src/config/test_config.py` in my PR branch by setting `ENABLE_BENCHMARK_DEV_MODE = True`
* When PR is reviewed and ready to be merged:
- [ ] I've reverted the code change on the config file mentioned above

## Reviewer Checklist
* For reviewer, before merging, please cross-check:
- [ ] I've verified the code change on the config file mentioned above has already been reverted

*Description:*

*Tests run:*

*DLC image/dockerfile:*

*Additional context:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/test_utils/__init__.py [hunk 1]
```diff
@@ -326,6 +328,7 @@ def get_inference_run_command(image_uri, model_names, processor="cpu"):
         multi_model_location = {
             "squeezenet": "https://torchserve.s3.amazonaws.com/mar_files/squeezenet1_1.mar",
             "pytorch-densenet": "https://torchserve.s3.amazonaws.com/mar_files/densenet161.mar",
+            "pytorch-resnet-neuron": "https://aws-dlc-sample-models.s3.amazonaws.com//pytorch//Resnet50-neuron.mar",
```
Comment (SergTogul) [inline test/test_utils/__init__.py:331]: Why do we use double slashes here?

Comment (aws-vrnatham) [reply test/test_utils/__init__.py:331]: Yes. It was typo mistake. Will fix it. Thanks

**Review by tejaschumbalkar — COMMENTED** [review]
LGTM. Will approve once the config change is reverted.

**Review by tejaschumbalkar — DISMISSED** [review]

File: pytorch/inference/docker/1.5.1/py3/Dockerfile.neuron [hunk 2]
```diff
@@ -1,4 +1,154 @@
 FROM ubuntu:16.04
 
-# Placeholder Dockerfile for Neuron
+LABEL dlc_major_version="1"
+LABEL maintainer="Amazon AI"
+LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
+LABEL com.amazonaws.sagemaker.capabilities.multi-models=true
```
Comment (saimidu) [inline pytorch/inference/docker/1.5.1/py3/Dockerfile.neuron:6]: The TF 1.15 dockerfile for Neuron support multi-model inference: https://github.com/aws/deep-learning-containers/blob/master/tensorflow/inference/docker/1.15/py3/Dockerfile.neuron

Please verify if this label is applicable here.

File: pytorch/buildspec-neuron.yml [hunk 3]
```diff
@@ -24,7 +27,7 @@
     BuildNeuronPTInferencePy3DockerImage:
       <<: *INFERENCE_REPOSITORY
       build: &PYTORCH_CPU_INFERENCE_PY3 false
-      image_size_baseline: 4899
+      image_size_baseline: 10000
```
Comment (saimidu) [inline pytorch/buildspec-neuron.yml:30]: What is the true size of the Neuron Inference DLC image?

File: pytorch/inference/docker/1.5.1/py3/Dockerfile.neuron [hunk 4]
```diff
@@ -1,4 +1,154 @@
...
+
+
+RUN wget https://www.open-mpi.org/software/ompi/v4.0/downloads/openmpi-$OPEN_MPI_VERSION.tar.gz \
+ && gunzip -c openmpi-$OPEN_MPI_VERSION.tar.gz | tar xf - \
+ && cd openmpi-$OPEN_MPI_VERSION \
+ && ./configure --prefix=/home/.openmpi \
+ && make all install \
+ && cd .. \
+ && rm openmpi-$OPEN_MPI_VERSION.tar.gz \
+ && rm -rf openmpi-$OPEN_MPI_VERSION
```
Comment (saimidu) [inline pytorch/inference/docker/1.5.1/py3/Dockerfile.neuron:80]: Does the Neuron Inference DLC require MPI?

File: test/dlc_tests/eks/pytorch/inference/test_eks_pytorch_inference.py [hunk 5]
```diff
@@ -2,18 +2,71 @@
...
 from invoke import run
 
 import test.test_utils.eks as eks_utils
 import test.test_utils as test_utils
 
+@pytest.mark.model("resnet")
+def test_eks_pytorch_neuron_inference(pytorch_inference, neuron_only):
+    server_type = test_utils.get_inference_server_type(pytorch_inference)
+    if "eia" in pytorch_inference or "neuron" not in pytorch_inference:
+        pytest.skip("Skipping EKS Neuron Test for EIA and Non Neuron Images")
```
Comment (saimidu) [inline test/dlc_tests/eks/pytorch/inference/test_eks_pytorch_inference.py:16]: Since the Neuron test isn't applicable to EI images, why not just make this if-condition conditioned on `"neuron" not in pytorch_inference`?

Also, please change the message printed out to say "EKS Neuron Inference test is not applicable to EI images" or something similar.

File: test/dlc_tests/eks/pytorch/inference/test_eks_pytorch_inference.py [hunk 6]
```diff
@@ -2,18 +2,71 @@
...
+        run(f"kubectl delete deployment {selector_name}")
+        run(f"kubectl delete service {selector_name}")
 
 @pytest.mark.model("densenet")
 def test_eks_pytorch_densenet_inference(pytorch_inference):
     server_type = test_utils.get_inference_server_type(pytorch_inference)
-    if "eia" in pytorch_inference:
-        pytest.skip("Skipping EKS Test for EIA")
+    if "eia" in pytorch_inference or "neuron" in pytorch_inference:
+        pytest.skip("Skipping EKS Test for EIA/neuron")
```
Comment (saimidu) [inline test/dlc_tests/eks/pytorch/inference/test_eks_pytorch_inference.py:69]: This message is incorrect, and could make reading the logs confusing.

I would recommend placing a separate if-condition above the `if "eia" in pytorch_inference` check, which can check for `"neuron" in pytorch_inference`, and print the message `"PyTorch Neuron Inference test has its own test function"` or something similar.

**Comment by saimidu** [comment]
@aws-vrnatham Could you please take a look at the comments in the review above?

File: pytorch/inference/docker/1.5.1/py3/Dockerfile.neuron (hunk 2 above)
Comment (aws-vrnatham) [reply pytorch/inference/docker/1.5.1/py3/Dockerfile.neuron]: Ok. This is not needed and hence removed it.

File: pytorch/buildspec-neuron.yml (hunk 3 above)
Comment (aws-vrnatham) [reply pytorch/buildspec-neuron.yml:30]: Around 8G now.

File: pytorch/inference/docker/1.5.1/py3/Dockerfile.neuron (hunk 4 above)
Comment (aws-vrnatham) [reply pytorch/inference/docker/1.5.1/py3/Dockerfile.neuron]: Not needed and removed it.

File: test/dlc_tests/eks/pytorch/inference/test_eks_pytorch_inference.py (hunk 5 above)
Comment (aws-vrnatham) [reply test/dlc_tests/eks/pytorch/inference/test_eks_pytorch_inference.py]: Done.

File: test/dlc_tests/eks/pytorch/inference/test_eks_pytorch_inference.py (hunk 6 above)
Comment (aws-vrnatham) [reply test/dlc_tests/eks/pytorch/inference/test_eks_pytorch_inference.py]: Done.

**Review by saimidu — APPROVED** [review]
